src/mym.cpp

```diff
@@ -1800,3 +1808,83 @@ mxArray* deserialize(const char* rpSerial, const size_t rlength) {
         mxFree(p_cmp);
     return p_res;
 }
+char *hex2char(char *original_val, const size_t char_length) {
+    const uint8_t *pnt = (uint8_t *)original_val;
+    uint8_t *result_pnt = new uint8_t[char_length*4];
+    unsigned int offset = 0;
+    for( unsigned int a = 0; a < char_length; ++a )
+    {
+        if     (pnt[a]<=0x7F) {
+            result_pnt[a+offset] = pnt[a];
+        }
+        else if(pnt[a]<=0x7FF) {
+            result_pnt[a+offset] = ((pnt[a]>>6) + 0xC0);
+            result_pnt[a+offset+1] = ((pnt[a] & 0x3F) + 0x80);
+            offset += 1;
+        }
+        else if(0xD800<=pnt[a] && pnt[a]<=0xDFFF) {
+            mexErrMsgIdAndTxt("DataJoint:Deserialization:UTF8",
+                "Invalid block of UTF8 detected.");
+        } //invalid block of utf8
+        else if(pnt[a]<=0xFFFF) {
+            result_pnt[a+offset] = ((pnt[a]>>12) + 0xE0);
+            result_pnt[a+offset+1] = (((pnt[a]>>6) & 0x3F) + 0x80);
+            result_pnt[a+offset+2] = ((pnt[a] & 0x3F) + 0x80);
+            offset += 2;
+        }
+        else if(pnt[a]<=0x10FFFF) {
+            result_pnt[a+offset] = ((pnt[a]>>18) + 0xF0);
+            result_pnt[a+offset+1] = (((pnt[a]>>12) & 0x3F) + 0x80);
+            result_pnt[a+offset+2] = (((pnt[a]>>6) & 0x3F) + 0x80);
+            result_pnt[a+offset+3] = ((pnt[a] & 0x3F) + 0x80);
+            offset += 3;
+        }
+    }
+    result_pnt[char_length+offset]= 0x00;
+    return (char *)result_pnt;
+}
+char *char2hex(char *original_val, const size_t vlength, const size_t char_length) {
+    unsigned int idx = 0;
+    unsigned int curr_length = 0;
+    unsigned char u0,u1,u2,u3;
+    uint8_t *result_pnt = new uint8_t[char_length];

```

Comment:
not sure about mex programming - do we need to use some matlab allocator to allow matlab to GC memory?

Reply:
I don't believe it applies here since in C++, only variables defined with `malloc` need to be explicitly freed. Since they are initially defined with `new`, they should be in the stack and destroyed once the function ends.

Reply:
result_pnt is returned to the caller for subsequent use.. 

(seen elsewhere here in ` mxCreateString(hex2char(row[j], p_lengths[j]));`, etc)

digging further:

https://www.mathworks.com/help/matlab/matlab_external/memory-management-issues.html

indicates there  is some matlab-specific allocator, and it's actually used elsewhere in the code. We should probably be consistent here

Reply:
Hmm. I agree on consistency and it is what I was going for. Regarding `char2hex`, seems that I mistakenly removed the `mxCalloc` assignment as it is manually freed in the caller function. I can update this. Regarding `hex2char`, we should be good there as I do not store the result to a variable and simply modify inline the `char` before passing it in to initialize `mxCreateString`.

Reply:
I have included preallocation to address this now.